lib/internal/modules/esm/public_loader_proxy.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+const { InternalWorker } = require('internal/worker');
+const path = require('path');
+const { deserialize, serialize } = require('v8');
+const { debug } = require('internal/modules/esm/debug');
+
+
+/**
+ * The communication vehicle constituting all memory shared between the main and the worker
+ */
+const commsChannel = new SharedArrayBuffer(2048);
+/**
+ * The lock/unlock segment of the shared memory. Atomics require this to be a Int32Array. This
+ * segment is used to tell the main to sleep when the worker is processing, and vice versa (for the
+ * worker to sleep whilst the main thread is processing).
+ * 0 → main sleeps
+ * 1 → worker sleeps
+ */
+const lock = new Int32Array(commsChannel, 0, 4);
+/**
+ * The request & response segment of the shared memory. TextEncoder/Decoder (needed to convert
+ * requests & responses into a format supported by the comms channel) reads and writes with
+ * Uint8Array.
+ */
+const requestResponseData = new Uint8Array(commsChannel, 4, 2044);
```

Comment:
What happens if the serialized data is bigger than 2044 bytes? Instead of shared memory, you could use `postMessage` + `receiveMessageOnPort` to synchronously pass data between the main thread and the worker.

Example: https://github.com/babel/babel/blob/586a7eec25381988b285b68340702ef3ecccbc31/eslint/babel-eslint-parser/src/client.cjs#L65-L84

Reply:
> What happens if the serialized data is bigger than 2044 bytes?

I imagine bad things.

> Instead of shared memory, you could use `postMessage` + `receiveMessageOnPort` to synchronously pass data between the main thread and the worker.

We tried that but found Atomics were faster than postMessage, creating a live-lock. I'm not sure how it's working in babel (we tried a min-repro and found the live-lock at the most basic level).

The run-time overhead of using shared memory with Atomics is significantly favourable to postMessage, so if we can use Atomics, that would seem to be much better and simpler.